rhombus/rhombus/scribblings/ref-rx-pattern.scrbl

```diff
@@ -0,0 +1,664 @@
+ implicitly prefixed with @rhombus(bof) for use with methods like
+ @rhombus(Regexp.match) (as opposed to @rhombus(Regexp.match_in)).
+
+@examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'^ "a"'.match_in("a")
+    rx'^ "a"'.match_in("xa")
+    rx'^ "a"'.match_in("x\na")
+    rx'bof "a"'.match_in("x\na")
+    rx'enable_newline: ^ "a"'.match_in("x\na")
+)
+
+}
+
+@doc(
+  rx.macro '$'
+  rx.macro 'eof'
+  rx.macro '$ $identifier: $pat'
+  rx.macro '$ $identifier'
+  rx.macro '$ $int'
+  rx.macro '$ $expr'
+){
+
+ The @rhombus($, ~at rhombus/rx) operator is overloaded for multiple
+ uses:
+
+@itemlist(
+
+ @item{When not followed by anything, @rhombus($, ~at rhombus/rx)
+  matches the end of input or the position before a newline, analogous to
+  the way that @rhombus(^, ~at rhombus/rx) matches the start of input or
+  the position after a newline. The @rhombus(eof, ~at rhombus/rx) operator
+  matches the end of input.
+
+  @examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'"a" $'.match_in("a")
+    rx'"a" $'.match_in("az")
+  )}
+
+ @item{When followed by an identifier and a @colon for a block
+  containing @rhombus(pat), @rhombus($, ~at rhombus/rx) creates a
+  @deftech{capture group}. The portion of input that is matched against
+  @rhombus(pat) is recorded and associated with the name
+  @rhombus(identifier). If the enclosing pattern uses @rhombus(pat) zero
+  or multiple times, then @rhombus(identifier) is associated to
+  @rhombus(#false) if the pattern is used zero times, or it is associated
+  to the latest match if used multiple times.
+
+  @examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'any ($m: any)'.match("ab")
+    rx'any ($m: any)'.match("ab")[#'m]
+    rx'any ($m: any)*'.match("a")
+  ~repl:
+    def rx'any ($m: any)' = "ab"
+    m
+  )}
+
+ @item{When followed by an identifier and no subsequent block, then
+  @rhombus($, ~at rhombus/rx) is either a @deftech{backreference} to a
+  named @tech{capture group}, or it is a @deftech{splice} of a regexp that
+  is bound to @rhombus(identifier).
+
+  The use of @rhombus($, ~at rhombus/rx) forms a backreference if
+  @rhombus(identifier) is associated to a capture group anywhere in the
+  enclosing pattern; the backreference matches input that is the same as
+  the most recent match for the capture group (and never matches if the
+  capture group does not yet have a match).
+
+  @examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'any ($m: any) $m'.match("abb")
+    rx'any ($m: any) $m'.match("abc")
+  )
+
+  When @rhombus($, ~at rhombus/rx) forms a splice, then a regular
+  expression is formed dynamically by merging the referenced regexp into
+  the enclosing pattern. (A limitation: both the merged regexp and
+  enclosing pattern must be free of backreferences, because backreferences
+  need to be converted from names to absolute positions eagerly.)
+
+  @examples(
+  ~eval: rx_eval
+  ~defn:
+    fun labeled(key) :: RX:
+      rx'$key ": " $name: .*'
```

Comment:
Adding parentheses to the example would make sense. Otherwise, I think our experience has been that it's better to write things more readably by convention than to impose constraints that are ad hoc in that way. Also, parentheses would be inconvenient when you want to bind a match result with `rx_in` as a binding form, such as writing `def rx'$m: pattern' = input` to bind `m` to the full match.